### PR TITLE
doc: changes to merge criteria (4 eyes principle)

### LINKED ISSUE #61358: Define guidelines for improved four eyes principle
Background for this was feedback we got from the community in the 'Meet the Maintainers' session during ZDS 2023.

Right now we require at least 2 approvals (4 eyes) for a pull request to be merged. In the Zephyr case, all eyes (submitter, approvers and merger) can be of the same organisation or team. A change that might seem harmless and if merged quickly to address an issue or add a feature without having being reviewed by a larger group of users might have negative effects and should be avoided.
Ideally we want at least one set of eyes looking at the changes from a different organisation, This for example could be the person merging the change, however, having reviews and approvals from other organisation will simplify things further and the merger + the approval of the assignee removes any ambiguity about the review.

We can further optimize this as we go, but at minimum we shall avoid the following:

- (a) Submitter, Approvers and Merger are from the same organisation

Additionally, the following should be considered: 

- (b) Changes to common and shared code shall always have reviews from different organisations (at least one review and approval from a different organisation as the submitter)
- (c) with changes limited only to platform code (driver, soc, boards), at least the merger shall be from a different organisation.
- ....

Consider and list other possible guidelines below...

### PR DESCRIPTION
**Pull request by nashif**
- doc: roles: remove reference to slide
- doc: release: update merge criteria

Fixes #61358

### CODE REVIEW

File: doc/project/project_roles.rst
Comment (erwango) [inline doc/project/project_roles.rst:319]: In fact this is "from one of the the impacted subsystems".

File: doc/project/project_roles.rst
Comment (erwango) [inline doc/project/project_roles.rst:321]: This is a bit unclear, should it be at least two approvals out of these three categories, or two approvals coming from one category is ok ?

File: doc/project/project_roles.rst
Comment (erwango) [inline doc/project/project_roles.rst:325]: I think it wouldn't hurt defining "common and shared code". Is that strictly equivalent to non vendor specific code ?

File: doc/project/project_roles.rst
Comment (erwango) [inline doc/project/project_roles.rst:331]: Is that:
-If all approvals of a change come from the same organisation, it can only be merged by a release engineer coming from another organisation.
-A change coming from one organisation needs at least an approval from another organisation to be merged

?

File: doc/project/project_roles.rst
Comment (erwango) [inline doc/project/project_roles.rst:333]: What about ?
A change coming from one organisation needs at least an approval from another organisation to be merged

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:319]: yes, I will change the language

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:321]: at least two approvals from all categories, I can make this clearer

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:325]: sure

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:323]: What does "on the organisation level" mean? 
That the code is considered common?
That the eyes need to be from "the organization"?

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:328]: Does this refer to common driver APIs? or to any driver, soc, or board related code?
(because if the 2nd, I do not expect people from other organizations to prioritize the time to review and understand code that is specific to somebody else's boards or drivers)

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:330]: > " Release engineers shall not merge code changes originating from the same"

Does this mean a) "Release engineers shall not merge code changes originating from *their* same organization"? or b) if both submitter and reviewer are from the same organization?

(in case of b) this would be a severe problem for quite a few areas where there just isn't so many involved developers. In some areas even getting two actual reviews is just plain difficult )

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:334]: Are this exceptions to be done by the release managers as they go?
or can release managers get into trouble if they merge something on good faith/under the understanding there is appropriate reviews?

**Review by aescolar — CHANGES_REQUESTED** [review]

**Review by carlescufi — DISMISSED** [review]

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:334]: no, driver, soc and board implementations, not APIs. addressed.

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst]: it means that we shall not have submitter, approvers and mergers from the same organisation. It is recognised that we have some areas (beside hardware support and drivers, etc.) where this might be a problem with reviews, but in this case the only change introduced is that the merger is from a different org.

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:321]: > Maintainers and collaborators of other areas.

Is this meant to say "Maintainers or collaborators of other areas affected by the PR" ?

I'm wondering if these 3 sentences would not be better rephrased overall to something like:

```
* Minimal of 2 approvals.
* PRs should be reviewed by at least a maintainer or collaborator of each affected area; 
  Unless the changes to a given area are considered trivial enough, in which case
  approvals by other affected subsystems maintainers/collaborators would suffice.
```

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:339]: Isn't "To be able to merge such changes, at least one review shall be from a different organisation."
fully redundant with the previous sentence?

File: doc/project/project_roles.rst
Comment (aescolar) [reply doc/project/project_roles.rst:341]: I meant to imply that we should be clear about who makes the exception, i.e.
"Release engineers may make exceptions for areas with contributions largely coming from one"..

By the way, largly => largely

Sidecomment:

> When in doubt, just ask someone else to merge something you are not sure about.

When in doubt typically the merger moves to the next PR ;)

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:347]: Should we say "Hotfixes can be merged at any time after CI has passed, and are excluded from most of the conditions listed above"?

**Review by aescolar — DISMISSED** [review]
To me, it is looking much better with this phrasing.
A few more minor things.

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:347]: yep

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:339]: it is saying it in a different way.

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:317]: This sentence now sounds a bit funny. Maybe just:
"Minimal of 2 approvals, including an approval by the designated assignee." (?)

File: doc/project/project_roles.rst
Comment (aescolar) [inline doc/project/project_roles.rst:341]: largly => largely

**Review by aescolar — COMMENTED** [review]
Just very minors at this point from me.

File: doc/project/project_roles.rst
Comment (fabiobaltieri) [inline doc/project/project_roles.rst:357]: I would not even bother listing the individual tests. It's a required check anyway, this list changes over time, it's probably out of date anyway. Would not miss it if you were to drop it, even just "All required checks are passing" should be more than enough, sounds like saying "required check are required" :-)

File: doc/project/project_roles.rst
Comment (erwango) [reply doc/project/project_roles.rst:333]: Agreed

File: doc/project/project_roles.rst
Comment (erwango) [inline doc/project/project_roles.rst:341]: s/reviewes/reviews

File: doc/project/project_roles.rst
Comment (nashif) [reply doc/project/project_roles.rst:357]: sure, not much related to this PR though, we can clean this up in a followup.

File: doc/project/project_roles.rst
Comment (kartben) [inline doc/project/project_roles.rst:346]:
```suggestion
* A minimum review period of 2 business days, 4 hours for trivial changes (see
```

**Review by henrikbrixandersen — APPROVED** [review]

**Review by simhein — APPROVED** [review]

File: doc/project/project_roles.rst
Comment (marc-hb) [inline doc/project/project_roles.rst:322]:
```suggestion
* Four eyes principle on the organisation level. We already require at least 2
  approvals (basic four eyeS principle), however, such reviews and approvals
```

File: doc/project/project_roles.rst
Comment (marc-hb) [inline doc/project/project_roles.rst:347]: I find these very short but off-topic...

**Review by kartben — APPROVED** [review]

**Pull request merged**
